### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by CSMastermind**

### CODE REVIEW

File: lib/api.js [hunk 1]
```diff
@@ -504,17 +383,28 @@ module.exports = function (PersistObjectTemplate, baseClassForPersist) {
          * @param {string} joinKey source table field name
          * @returns {*[]}
          */
-        template.knexChildJoin = function (targetTemplate, primaryAlias, targetAlias, joinKey) {
+        template.persistorKnexChildJoin = function (targetTemplate, primaryAlias, targetAlias, joinKey) {
             return [template.getTableName() + ' as ' + primaryAlias, targetTemplate.getChildKey(joinKey, primaryAlias), targetTemplate.getPrimaryKey(targetAlias)];
         };
 
+
+        // Depericated API
```
Comment (hackerrdave) [inline lib/api.js]: "Deprecated"

File: lib/api.js [hunk 2]
```diff
@@ -349,20 +176,27 @@ module.exports = function (PersistObjectTemplate, baseClassForPersist) {
          * @param {object} txn persistObjectTemplate transaciton object
          * @param {object} logger objecttemplate logger
          * @returns {object}
+         * @deprecated in favor of persitorDeleteByQuery
          */
-        template.deleteFromPersistWithQuery = function(query, txn, logger) {
+        template.deleteFromPersistWithQuery = function(query, txn, logger) { //
```
Comment (hackerrdave) [inline lib/api.js]: empty comment?

File: lib/index.js [hunk 3]
```diff
@@ -50,8 +50,10 @@
  * @param baseClassForPersist
  */
 var nextId = 1;
+var objectTemplate;
+var supertype = require('supertype');
 
-module.exports = function (_ObjectTemplate, _RemoteObjectTemplate, baseClassForPersist) { //@TODO: Why is ObjectTemplate and RemoteObjectTemplate here?
+module.exports = function (ObjectTemplate, _RemoteObjectTemplate, baseClassForPersist) { //@TODO: Why is ObjectTemplate and RemoteObjectTemplate here?
```
Comment (hackerrdave) [inline lib/index.js]: is `ObjectTemplate` being used now?

File: lib/index.js [hunk 4]
```diff
@@ -71,5 +73,59 @@ module.exports = function (_ObjectTemplate, _RemoteObjectTemplate, baseClassForP
...
+        __extends(class_1, _super);
+        function class_1() {
+            return _super !== null && _super.apply(this, arguments) || this;
+        }
+        return class_1;
+    }(Base));
+}
+
+module.exports.Persistor = {
+    create: function () {return  module.exports(require('supertype'), null, require('supertype'))}
```
Comment (hackerrdave) [inline lib/index.js]: `module.exports` is an object, right? how does `module.exports()` work?

File: lib/index.js [hunk 5]
```diff
@@ -71,5 +73,59 @@ module.exports = function (_ObjectTemplate, _RemoteObjectTemplate, baseClassForP
...
+}
+
+Object.defineProperty(module.exports.Persistable.prototype, 'persistor', {get: function () {
+    return this.__objectTemplate__
+}});
+
+
+module.exports.PersistObjectTemplate = module.exports;
```
Comment (hackerrdave) [inline lib/index.js]: why making this assignment? isn't this basically duplicating the whole `module.exports` object?

File: test/persist_schema_updates.js [hunk 6]
```diff
@@ -401,8 +401,10 @@ describe('schema update checks', function () {
 
     it('without defining the default db alias', function () {
         var WithOutSchema = PersistObjectTemplate.create('WithOutSchema', {});
+        PersistObjectTemplate._injectObjectFunctions(WithOutSchema);
         var obj = new WithOutSchema();
-        expect(obj.persistSave.bind(this)).to.throw('DB Alias __default__ not set');
+        //    obj.persistSave();
```
Comment (hackerrdave) [inline test/persist_schema_updates.js]: delete commented code

File: lib/api.js (hunk 1 above)
Comment (soitof) [inline lib/api.js]: Depericate!

File: lib/index.js (hunk 4 above)
Comment (soitof) [inline lib/index.js]: That would invoke whatever the module exports.

Comment (selsamman) [inline lib/index.js]: module.exports is persistor (PersistObjectTemplate)

File: lib/index.js (hunk 5 above)
Comment (selsamman) [inline lib/index.js]: That is a head scratcher.  I can't see anyone referencing persistor this way.  I tried taking it out and all persistor and amorphic tests still run.  Must have been some idea that was later superseded by something else.  Feel free to ditch it.

File: lib/api.js (hunk 2 above)
Comment (CSMastermind) [inline lib/api.js]: Fixed

File: lib/api.js (hunk 1 above)
Comment (CSMastermind) [inline lib/api.js]: Decapitated?

Comment (CSMastermind) [inline lib/api.js]: Fixed.

File: lib/index.js (hunk 3 above)
Comment (CSMastermind) [inline lib/index.js]: It is not (fixed).

File: lib/index.js (hunk 5 above)
Comment (CSMastermind) [inline lib/index.js]: Went ahead and pulled it out.

File: test/persist_schema_updates.js (hunk 6 above)
Comment (CSMastermind) [inline test/persist_schema_updates.js]: Fixed

**Pull request merged**